// @ghostery/urlfilter2dnr/dist/esm/converters/adguard.js
import { DeclarativeFilterConverter, } from '@adguard/tsurlfilter/es/declarative-converter';
import { normalizeFilter, normalizeRule } from './helpers.js';
const converter = new DeclarativeFilterConverter();
class SimpleFilterList {
    content;
    constructor(content) {
        this.content = content;
    }
    getContent() {
        return this.content;
    }
    getConversionData() {
        return { originals: [], conversions: {} };
    }
    getRuleText(offset) {
        if (offset >= this.content.length)
            return null;
        const nlIndex = this.content.indexOf('\n', offset);
        const endIndex = nlIndex === -1 ? this.content.length : nlIndex;
        const line = this.content.slice(offset, endIndex);
        return line.endsWith('\r') ? line.slice(0, -1) : line;
    }
    getOriginalRuleText(offset) {
        if (offset < 0 || offset >= this.content.length)
            return null;
        return this.getRuleText(offset);
    }
    getConvertedRuleOriginal() {
        return null;
    }
    getOriginalContent() {
        return this.content;
    }
}
const createFilter = (rules, filterId = 0) => {
    const filterList = new SimpleFilterList(rules.join('\n'));
    return {
        getId: () => filterId,
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        getContent: async () => filterList,
        getRuleByIndex: async (index) => filterList.getOriginalRuleText(index) ?? '',
        isTrusted: () => true,
        unloadContent: () => { },
    };
};
export default async function convert(rules, { resourcesPath = '/prefix' } = {}) {
    if (rules.length === 0) {
        return {
            rules: [],
            errors: [],
            limitations: [],
        };
    }
    const filter = createFilter(rules.map((rule) => normalizeFilter(rule) ?? ''));
    const conversionResult = await converter.convertStaticRuleSet(filter, { resourcesPath });
    const declarativeRules = await conversionResult.ruleSet.getDeclarativeRules();
    const normalizeRules = [];
    const errors = conversionResult.errors.map((e) => e.toString());
    for (const [index, rule] of declarativeRules.entries()) {
        try {
            normalizeRules.push(normalizeRule(rule, { resourcesPath, id: index + 1 }));
        }
        catch (e) {
            errors.push(`Could not normalize rule: ${JSON.stringify(rule)} - ${e instanceof Error ? e.message : e}`);
        }
    }
    return {
        rules: normalizeRules,
        errors,
        limitations: conversionResult.limitations,
    };
}
