// @adguard/agtree/dist/utils/index.js
/*
 * AGTree v4.1.1 (build date: Thu, 23 Apr 2026 09:15:37 GMT)
 * (c) 2026 Adguard Software Ltd.
 * Released under the MIT license
 * https://github.com/AdguardTeam/tsurlfilter/tree/master/packages/agtree#readme
 */
export { ADG_SCRIPTLET_MASK, AGLINT_COMMAND_PREFIX, COMMA, COMMA_DOMAIN_LIST_SEPARATOR, HINT_MARKER, IF, INCLUDE, MODIFIERS_SEPARATOR, MODIFIER_ASSIGN_OPERATOR, NEGATION_MARKER, NETWORK_RULE_EXCEPTION_MARKER, NETWORK_RULE_SEPARATOR, PIPE, PIPE_MODIFIER_SEPARATOR, PREPROCESSOR_MARKER, SAFARI_CB_AFFINITY, SLASH, UBO_SCRIPTLET_MASK } from './constants.js';
export { AdblockSyntax } from './adblockers.js';
export { getBitCount } from './bit-count.js';
export { CosmeticRuleSeparatorUtils } from './cosmetic-rule-separator.js';
export { DomainUtils } from './domain.js';
export { LogicalExpressionUtils } from './logical-expression.js';
export { ADBLOCK_URL_SEPARATOR, ADBLOCK_URL_SEPARATOR_REGEX, ADBLOCK_URL_START, ADBLOCK_URL_START_REGEX, ADBLOCK_WILDCARD, ADBLOCK_WILDCARD_REGEX, RegExpUtils, SPECIAL_REGEX_SYMBOLS } from './regexp.js';
export { QuoteType, QuoteUtils } from './quotes.js';
export { PositionProvider } from './position-provider.js';
export { RuleCategorizer } from './categorizer.js';
