// @adguard/tsurlfilter/dist/es/declarative-converter-utils.js
import { z } from 'zod';

var ErrorStatusCodes;
(function (ErrorStatusCodes) {
    ErrorStatusCodes[ErrorStatusCodes["ComplexRegex"] = 1001] = "ComplexRegex";
    ErrorStatusCodes[ErrorStatusCodes["RuleLimit"] = 1002] = "RuleLimit";
    ErrorStatusCodes[ErrorStatusCodes["RegexpRuleLimit"] = 1003] = "RegexpRuleLimit";
    ErrorStatusCodes[ErrorStatusCodes["RemoveparamRegexpIsNotSupported"] = 1004] = "RemoveparamRegexpIsNotSupported";
    ErrorStatusCodes[ErrorStatusCodes["RemoveparamInversionIsNotSupported"] = 1005] = "RemoveparamInversionIsNotSupported";
})(ErrorStatusCodes || (ErrorStatusCodes = {}));
const EMPTY_STRING = '';
const RE_NUMBER = /^\d+$/;
/**
 * Prefix for ruleset name.
 */
const RULESET_NAME_PREFIX = 'ruleset_';

/**
 * Extension for ruleset file.
 */
const RULESET_FILE_EXT = '.json';
/**
 * Path separator.
 */
const PATH_SEPARATOR = '/';
/**
 * Helper method to get the rule set ID with the {@link RULESET_NAME_PREFIX} prefix.
 *
 * @param ruleSetId Rule set id. Can be a number or a string.
 *
 * @returns Rule set ID with the {@link RULESET_NAME_PREFIX} prefix.
 */
function getRuleSetId(ruleSetId) {
    let ruleSetIdStr = String(ruleSetId);
    if (!ruleSetIdStr.startsWith(RULESET_NAME_PREFIX)) {
        ruleSetIdStr = `${RULESET_NAME_PREFIX}${ruleSetIdStr}`;
    }
    return ruleSetIdStr;
}
/**
 * Helper method to get the path to the rule set file.
 *
 * @param ruleSetId Rule set id. Can be a number or a string.
 * @param baseDir Base directory.
 *
 * @returns Path to the rule set file.
 *
 * @note This is just a path, not a URL. To get a URL, use `browser.runtime.getURL`.
 * @note Rule set ID automatically gets a {@link RULESET_NAME_PREFIX} prefix if it doesn't have it,
 * e.g. `123` -> `ruleset_123` or `foo` -> `ruleset_foo`.
 */
function getRuleSetPath(ruleSetId, baseDir) {
    const ruleSetIdWithPrefix = getRuleSetId(ruleSetId);
    const resultPrefix = baseDir ? `${baseDir}${PATH_SEPARATOR}` : EMPTY_STRING;
    return `${resultPrefix}${ruleSetIdWithPrefix}${PATH_SEPARATOR}${ruleSetIdWithPrefix}${RULESET_FILE_EXT}`;
}
/**
 * Helper method to extract the rule set ID from the given string.
 *
 * This method processes a string that may represent a path or contain a prefix,
 * extracts the last part if it contains slashes, removes a defined prefix if present,
 * and validates whether the remaining part is a numeric string.
 *
 * @param ruleSetId The rule set ID or path to process.
 *
 * @returns The extracted rule set ID as a number, or `null` if the ID cannot be extracted.
 */
function extractRuleSetId(ruleSetId) {
    let ruleSetIdToParse = ruleSetId;
    // Extract the last part of the path if it contains '/'
    const lastSlashIndex = ruleSetIdToParse.lastIndexOf(PATH_SEPARATOR);
    if (lastSlashIndex !== -1) {
        ruleSetIdToParse = ruleSetIdToParse.slice(lastSlashIndex + PATH_SEPARATOR.length);
    }
    // Remove the prefix and file extension if present
    if (ruleSetIdToParse.startsWith(RULESET_NAME_PREFIX)) {
        ruleSetIdToParse = ruleSetIdToParse.slice(RULESET_NAME_PREFIX.length);
        if (ruleSetIdToParse.endsWith(RULESET_FILE_EXT)) {
            ruleSetIdToParse = ruleSetIdToParse.slice(0, -RULESET_FILE_EXT.length);
        }
    }
    // Validate and parse as a number
    if (!RE_NUMBER.test(ruleSetIdToParse)) {
        return null;
    }
    const possibleInt = parseInt(ruleSetIdToParse, 10);
    return Number.isNaN(possibleInt) ? null : possibleInt;
}

/**
 * RequestType is the request types enumeration.
 * Important: the enumeration is marked as const to avoid side effects when
 * importing it into an extension.
 */
const RequestType = {
    /**
     * Main frame.
     */
    Document: 1,
    /**
     * (iframe) $subdocument.
     */
    SubDocument: 2, // 1 << 1
    /**
     * (javascript, etc) $script.
     */
    Script: 4, // 1 << 2
    /**
     * (css) $stylesheet.
     */
    Stylesheet: 8, // 1 << 3
    /**
     * (flash, etc) $object.
     */
    Object: 16, // 1 << 4
    /**
     * (any image) $image.
     */
    Image: 32, // 1 << 5
    /**
     * (ajax/fetch) $xmlhttprequest.
     */
    XmlHttpRequest: 64, // 1 << 6
    /**
     * (video/music) $media.
     */
    Media: 128, // 1 << 7
    /**
     * (any custom font) $font.
     */
    Font: 256, // 1 << 8
    /**
     * (a websocket connection) $websocket.
     */
    WebSocket: 512, // 1 << 9
    /**
     * (navigator.sendBeacon()) $ping.
     */
    Ping: 1024, // 1 << 10
    /**
     * Any other request type.
     */
    Other: 4096, // 1 << 12
};

var HTTPMethod;
(function (HTTPMethod) {
    HTTPMethod["GET"] = "GET";
    HTTPMethod["POST"] = "POST";
    HTTPMethod["PUT"] = "PUT";
    HTTPMethod["DELETE"] = "DELETE";
    HTTPMethod["PATCH"] = "PATCH";
    HTTPMethod["HEAD"] = "HEAD";
    HTTPMethod["OPTIONS"] = "OPTIONS";
    HTTPMethod["CONNECT"] = "CONNECT";
    HTTPMethod["TRACE"] = "TRACE";
})(HTTPMethod || (HTTPMethod = {}));

/* eslint-disable jsdoc/require-description-complete-sentence */
/**
 * @file Describes types from declarativeNetRequest,
 * since @types/chrome does not contain actual information.
 *
 * Updated 07/09/2022.
 */
/**
 * https://developer.chrome.com/docs/extensions/reference/declarativeNetRequest/#type-DomainType
 */
var DomainType;
(function (DomainType) {
    DomainType["FirstParty"] = "firstParty";
    DomainType["ThirdParty"] = "thirdParty";
})(DomainType || (DomainType = {}));
/**
 * https://developer.chrome.com/docs/extensions/reference/declarativeNetRequest/#type-ResourceType
 */
var ResourceType;
(function (ResourceType) {
    ResourceType["MainFrame"] = "main_frame";
    ResourceType["SubFrame"] = "sub_frame";
    ResourceType["Stylesheet"] = "stylesheet";
    ResourceType["Script"] = "script";
    ResourceType["Image"] = "image";
    ResourceType["Font"] = "font";
    ResourceType["Object"] = "object";
    ResourceType["XmlHttpRequest"] = "xmlhttprequest";
    ResourceType["Ping"] = "ping";
    ResourceType["Media"] = "media";
    ResourceType["WebSocket"] = "websocket";
    ResourceType["Other"] = "other";
    // Temporary not using
    // TODO: Add csp_report handler similar to AG-24613 but in declarative way.
    // CspReport = 'csp_report',
    // WebTransport = 'webtransport',
    // WebBundle = 'webbundle',
})(ResourceType || (ResourceType = {}));
/**
 * https://developer.chrome.com/docs/extensions/reference/declarativeNetRequest/#type-QueryKeyValue
 */
const QueryKeyValueValidator = z.strictObject({
    key: z.string(),
    replaceOnly: z.boolean().optional(),
    value: z.string(),
});
/**
 * https://developer.chrome.com/docs/extensions/reference/declarativeNetRequest/#type-QueryTransform
 */
const QueryTransformValidator = z.strictObject({
    addOrReplaceParams: QueryKeyValueValidator.array().optional(),
    removeParams: z.string().array().optional(),
});
/**
 * https://developer.chrome.com/docs/extensions/reference/declarativeNetRequest/#type-URLTransform
 */
const URLTransformValidator = z.strictObject({
    fragment: z.string().optional(),
    host: z.string().optional(),
    password: z.string().optional(),
    path: z.string().optional(),
    port: z.string().optional(),
    query: z.string().optional(),
    queryTransform: QueryTransformValidator.optional(),
    scheme: z.string().optional(),
    username: z.string().optional(),
});
/**
 * https://developer.chrome.com/docs/extensions/reference/declarativeNetRequest/#type-Redirect
 */
const RedirectValidator = z.strictObject({
    extensionPath: z.string().optional(),
    regexSubstitution: z.string().optional(),
    transform: URLTransformValidator.optional(),
    url: z.string().optional(),
});
/**
 * https://developer.chrome.com/docs/extensions/reference/declarativeNetRequest/#type-HeaderOperation
 */
var HeaderOperation;
(function (HeaderOperation) {
    HeaderOperation["Append"] = "append";
    HeaderOperation["Set"] = "set";
    HeaderOperation["Remove"] = "remove";
})(HeaderOperation || (HeaderOperation = {}));
/**
 * https://developer.chrome.com/docs/extensions/reference/declarativeNetRequest/#type-ModifyHeaderInfo
 */
const ModifyHeaderInfoValidator = z.strictObject({
    header: z.string(),
    operation: z.nativeEnum(HeaderOperation),
    value: z.string().optional(),
});
/**
 * https://developer.chrome.com/docs/extensions/reference/declarativeNetRequest/#type-RuleActionType
 */
var RuleActionType;
(function (RuleActionType) {
    RuleActionType["BLOCK"] = "block";
    RuleActionType["REDIRECT"] = "redirect";
    RuleActionType["ALLOW"] = "allow";
    RuleActionType["UPGRADE_SCHEME"] = "upgradeScheme";
    RuleActionType["MODIFY_HEADERS"] = "modifyHeaders";
    /**
     * For allowAllRequests rules {@link RuleCondition.resourceTypes} may only
     * include the 'sub_frame' and 'main_frame' resource types.
     */
    RuleActionType["ALLOW_ALL_REQUESTS"] = "allowAllRequests";
})(RuleActionType || (RuleActionType = {}));
/**
 * https://developer.chrome.com/docs/extensions/reference/declarativeNetRequest/#type-RuleAction
 */
const RuleActionValidator = z.strictObject({
    redirect: RedirectValidator.optional(),
    requestHeaders: ModifyHeaderInfoValidator.array().optional(),
    responseHeaders: ModifyHeaderInfoValidator.array().optional(),
    type: z.nativeEnum(RuleActionType),
});
/**
 * https://developer.chrome.com/docs/extensions/reference/declarativeNetRequest/#type-RequestMethod
 */
var RequestMethod;
(function (RequestMethod) {
    RequestMethod["Connect"] = "connect";
    RequestMethod["Delete"] = "delete";
    RequestMethod["Get"] = "get";
    RequestMethod["Head"] = "head";
    RequestMethod["Options"] = "options";
    RequestMethod["Patch"] = "patch";
    RequestMethod["Post"] = "post";
    RequestMethod["Put"] = "put";
})(RequestMethod || (RequestMethod = {}));
/**
 * Map {@link HTTPMethod} to declarative {@link RequestMethod}.
 */
({
    [HTTPMethod.GET]: RequestMethod.Get,
    [HTTPMethod.POST]: RequestMethod.Post,
    [HTTPMethod.PUT]: RequestMethod.Put,
    [HTTPMethod.DELETE]: RequestMethod.Delete,
    [HTTPMethod.PATCH]: RequestMethod.Patch,
    [HTTPMethod.HEAD]: RequestMethod.Head,
    [HTTPMethod.OPTIONS]: RequestMethod.Options,
    [HTTPMethod.CONNECT]: RequestMethod.Connect,
});
/**
 * https://developer.chrome.com/docs/extensions/reference/declarativeNetRequest/#type-HeaderInfo
 * Chrome 128+
 */
const HeaderInfoValidator = z.strictObject({
    excludedValues: z.string().array().optional(),
    header: z.string(),
    values: z.string().array().optional(),
});
/**
 * https://developer.chrome.com/docs/extensions/reference/declarativeNetRequest/#type-RuleCondition
 */
const RuleConditionValidator = z.strictObject({
    domainType: z.nativeEnum(DomainType).optional(),
    excludedInitiatorDomains: z.string().array().optional(),
    excludedRequestDomains: z.string().array().optional(),
    excludedRequestMethods: z.nativeEnum(RequestMethod).array().optional(),
    excludedResourceTypes: z.nativeEnum(ResourceType).array().optional(),
    excludedTabIds: z.number().array().optional(),
    initiatorDomains: z.string().array().optional(),
    isUrlFilterCaseSensitive: z.boolean().optional(),
    regexFilter: z.string().optional(),
    requestDomains: z.string().array().optional(),
    requestMethods: z.nativeEnum(RequestMethod).array().optional(),
    /**
     * Rule matches if the request matches any response header condition in this list (if specified).
     * Chrome 128+
     */
    responseHeaders: HeaderInfoValidator.array().optional(),
    /**
     * If none of the `excludedResourceTypes` and `resourceTypes` are specified,
     * all resource types except "main_frame" will be matched.
     */
    resourceTypes: z.nativeEnum(ResourceType).array().optional(),
    tabIds: z.number().array().optional(),
    urlFilter: z.string().optional(),
});
/**
 * https://developer.chrome.com/docs/extensions/reference/declarativeNetRequest/#type-Rule
 */
z.strictObject({
    action: RuleActionValidator,
    condition: RuleConditionValidator,
    id: z.number(),
    priority: z.number().optional(),
});
/**
 * Map request types to declarative types.
 */
({
    [ResourceType.MainFrame]: RequestType.Document,
    [ResourceType.SubFrame]: RequestType.SubDocument,
    [ResourceType.Stylesheet]: RequestType.Stylesheet,
    [ResourceType.Script]: RequestType.Script,
    [ResourceType.Image]: RequestType.Image,
    [ResourceType.Font]: RequestType.Font,
    [ResourceType.Object]: RequestType.Object,
    [ResourceType.XmlHttpRequest]: RequestType.XmlHttpRequest,
    [ResourceType.Ping]: RequestType.Ping,
    // TODO: what should match this resource type?
    // [ResourceType.CSP_REPORT]: RequestType.Document,
    [ResourceType.Media]: RequestType.Media,
    [ResourceType.WebSocket]: RequestType.WebSocket,
    [ResourceType.Other]: RequestType.Other,
});

/**
 * @file Separate file for the `isSafeRule` function, because it can be used
 * outside of the declarative converter to check if a rule is safe, e.g. for
 * passing rules to "skip review" in CWS.
 */
/**
 * List of declarative rule actions which are considered safe.
 *
 * @see {@link https://developer.chrome.com/docs/extensions/reference/api/declarativeNetRequest#safe_rules}
 */
const SAFE_RULE_ACTIONS = new Set([
    RuleActionType.BLOCK,
    RuleActionType.ALLOW,
    RuleActionType.ALLOW_ALL_REQUESTS,
    RuleActionType.UPGRADE_SCHEME,
]);
/**
 * Checks whether the declarative rule is safe.
 *
 * @see {@link https://developer.chrome.com/docs/extensions/reference/api/declarativeNetRequest#safe_rules}
 *
 * @param rule Declarative rule to check.
 *
 * @returns True if the rule is safe, otherwise false.
 */
const isSafeRule = (rule) => {
    return SAFE_RULE_ACTIONS.has(rule.action.type);
};

export { RULESET_FILE_EXT, extractRuleSetId, getRuleSetId, getRuleSetPath, isSafeRule };
