// @adguard/agtree/dist/index.js
/*
 * AGTree v4.1.1 (build date: Thu, 23 Apr 2026 09:15:37 GMT)
 * (c) 2026 Adguard Software Ltd.
 * Released under the MIT license
 * https://github.com/AdguardTeam/tsurlfilter/tree/master/packages/agtree#readme
 */
export { RuleParser } from './parser/rule-parser.js';
export { RuleGenerator } from './generator/rule-generator.js';
export { defaultParserOptions } from './parser/options.js';
export { CommentMarker, CommentRuleType, CosmeticRuleSeparator, CosmeticRuleType, NetworkRuleType, RuleCategory } from './nodes/index.js';
export { AdblockSyntaxError } from './errors/adblock-syntax-error.js';
export { AgentCommentParser } from './parser/comment/agent-comment-parser.js';
export { AgentParser } from './parser/comment/agent-parser.js';
export { CommentParser } from './parser/comment/comment-parser.js';
export { ConfigCommentParser } from './parser/comment/config-comment-parser.js';
export { CosmeticRuleParser } from './parser/cosmetic/cosmetic-rule-parser.js';
export { AppListParser } from './parser/misc/app-list-parser.js';
export { DomainListParser } from './parser/misc/domain-list-parser.js';
export { MethodListParser } from './parser/misc/method-list-parser.js';
export { StealthOptionListParser } from './parser/misc/stealth-option-list-parser.js';
export { FilterListParser } from './parser/filterlist-parser.js';
export { HintCommentParser } from './parser/comment/hint-comment-parser.js';
export { HintParser } from './parser/comment/hint-parser.js';
export { LogicalExpressionParser } from './parser/misc/logical-expression-parser.js';
export { MetadataCommentParser } from './parser/comment/metadata-comment-parser.js';
export { ModifierListParser } from './parser/misc/modifier-list.js';
export { ModifierParser } from './parser/misc/modifier-parser.js';
export { NetworkRuleParser } from './parser/network/network-rule-parser.js';
export { NotImplementedError } from './errors/not-implemented-error.js';
export { ParameterListParser } from './parser/misc/parameter-list-parser.js';
export { HostRuleParser } from './parser/network/host-rule-parser.js';
export { PreProcessorCommentParser } from './parser/comment/preprocessor-parser.js';
export { RuleConversionError } from './errors/rule-conversion-error.js';
export { modifierValidator } from './validator/index.js';
export { FilterListConverter } from './converter/filter-list.js';
export { RawFilterListConverter } from './converter/raw-filter-list.js';
export { RuleConverter } from './converter/rule.js';
export { RawRuleConverter } from './converter/raw-rule.js';
export { ADG_SCRIPTLET_MASK, AGLINT_COMMAND_PREFIX, COMMA_DOMAIN_LIST_SEPARATOR, HINT_MARKER, IF, INCLUDE, MODIFIERS_SEPARATOR, MODIFIER_ASSIGN_OPERATOR, NEGATION_MARKER, NETWORK_RULE_EXCEPTION_MARKER, NETWORK_RULE_SEPARATOR, PIPE_MODIFIER_SEPARATOR, PREPROCESSOR_MARKER, SAFARI_CB_AFFINITY, UBO_SCRIPTLET_MASK } from './utils/constants.js';
export { AdblockProduct, AdblockSyntax, getHumanReadableProductName } from './utils/adblockers.js';
export { CosmeticRuleSeparatorUtils } from './utils/cosmetic-rule-separator.js';
export { DomainUtils } from './utils/domain.js';
export { LogicalExpressionUtils } from './utils/logical-expression.js';
export { ADBLOCK_URL_SEPARATOR, ADBLOCK_URL_SEPARATOR_REGEX, ADBLOCK_URL_START, ADBLOCK_URL_START_REGEX, ADBLOCK_WILDCARD, ADBLOCK_WILDCARD_REGEX, RegExpUtils, SPECIAL_REGEX_SYMBOLS } from './utils/regexp.js';
export { QuoteType, QuoteUtils } from './utils/quotes.js';
export { PositionProvider } from './utils/position-provider.js';
export { RuleCategorizer } from './utils/categorizer.js';
export { NATIVE_CSS_PSEUDO_CLASSES, REMOVE_PROPERTY } from './converter/data/css.js';
export { CssTokenStream } from './parser/css/css-token-stream.js';
export { modifiersCompatibilityTable } from './compatibility-tables/modifiers.js';
export { redirectsCompatibilityTable } from './compatibility-tables/redirects.js';
export { scriptletsCompatibilityTable } from './compatibility-tables/scriptlets.js';
export { GenericPlatform, SpecificPlatform } from './compatibility-tables/platforms.js';
import './compatibility-tables/schemas/base.js';
import './compatibility-tables/schemas/modifier.js';
import './compatibility-tables/schemas/redirect.js';
import './compatibility-tables/schemas/scriptlet.js';
export { PLATFORM_NEGATION, PLATFORM_SEPARATOR, parseRawPlatforms, stringifyPlatforms } from './compatibility-tables/schemas/platform.js';
export { getAllPlatformNames, getHumanReadablePlatformName, getPlatformId, getPlatformsByProduct, getProductGenericPlatforms, getProductSpecificPlatforms, getSpecificPlatformName, hasPlatformMultipleProducts, isGenericPlatform, platformToAdblockProduct } from './compatibility-tables/utils/platform-helpers.js';
export { ResourceType } from './compatibility-tables/schemas/resource-type.js';
export { getResourceTypeModifier, isValidResourceType } from './compatibility-tables/utils/resource-type-helpers.js';
export { AGTREE_VERSION } from './version.js';
